File: apps/api/app/ratelimit.py
"""Redis-backed rate limiting for the search endpoint.

Fixed-window counter per client IP (one Redis key per minute, reusing the
answer-cache client). Fails open: if Redis is unreachable we log a warning and
allow the request, so a cache outage never takes the API down with it.
"""

from __future__ import annotations

import logging
import time

from . import cache
from .config import settings

log = logging.getLogger("yunoball.ratelimit")

WINDOW_SECONDS = 60


async def retry_after(client_ip: str) -> int | None:
    """Count this request against `client_ip`'s current window.

    Returns None to allow the request, or the number of seconds until the
    window resets when the client exceeded `settings.rate_limit_per_minute`
    (0 disables limiting entirely).
    """
    limit = settings.rate_limit_per_minute
    if limit <= 0:
        return None
    now = int(time.time())
    key = f"yb:ratelimit:{client_ip}:{now // WINDOW_SECONDS}"
    try:
        client = cache.get_client()
        count = await client.incr(key)
        if count == 1:
            await client.expire(key, WINDOW_SECONDS)
    except Exception:  # noqa: BLE001
        log.warning("rate limit check failed (Redis unreachable); allowing request")
        return None
    if count > limit:
        return WINDOW_SECONDS - now % WINDOW_SECONDS
    return None

File: apps/server/src/lib/rateLimit.ts
/** Fixed-window rate limiting per client IP for the search/agent endpoints.
 * In-process counters (single-instance API); windows expire by key rotation.
 * ponytail: swap the Map for Redis INCR if the API ever runs multi-instance. */

import type { Request } from "express";
import { config } from "../config.js";

const WINDOW_SECONDS = 60;
const counts = new Map<string, number>();
let currentWindow = 0;

// Cap on distinct keys held in a single window — a spoofed/rotated source can
// otherwise insert unbounded entries before the next-window clear().
const MAX_KEYS = 100_000;

export function clientIp(req: Request): string {
  // req.ip honors the app's `trust proxy` setting (configured in app.ts), so a
  // client can't mint a fresh bucket by forging X-Forwarded-For — only the
  // configured number of proxy hops are trusted.
  return req.ip ?? req.socket.remoteAddress ?? "unknown";
}

/** Returns null to allow, or seconds until the window resets when over limit.
 * `key` should namespace the tier (e.g. "w:<ip>" vs "r:<ip>") so read and write
 * tiers count in separate buckets; `limit` defaults to the write limit. */
export function retryAfter(key: string, limit = config.rateLimitPerMinute): number | null {
  if (limit <= 0) return null;
  const now = Math.floor(Date.now() / 1000);
  const window = Math.floor(now / WINDOW_SECONDS);
  if (window !== currentWindow) {
    counts.clear();
    currentWindow = window;
  }
  const n = (counts.get(key) ?? 0) + 1;
  // Stop tracking new keys once the map is saturated within a window; existing
  // keys keep counting so real clients are still limited.
  if (counts.has(key) || counts.size < MAX_KEYS) counts.set(key, n);
  if (n > limit) return WINDOW_SECONDS - (now % WINDOW_SECONDS);
  return null;
}
